src/room007/models/word_tag_predictor.py

```diff
@@ -15,117 +17,130 @@
 from sklearn.linear_model import LogisticRegression
 
 from room007.models import model
+from room007.data import info
 
+import spacy
+#nlp = spacy.load('en')
 
 stop_words = nltk.corpus.stopwords.words('english') + [x for x in string.printable]
 
-
-class Features(object):
+@model.FeatureManager
+def _is_in_question(data):
+    def is_in_question(row):
+        features = []
+        question = 0
+        for word in row.split()[::-1]:
+            if word in '.:;!?':
+                question = int(word == '?')
+            if word not in stop_words:
+                features.append(question)
+        return features[::-1]
+    return data['titlecontent'].apply(
+        is_in_question
+        )
+
+@model.FeatureManager
+def _title_or_content(data):
+    return data.apply(
+        lambda row: [1] * len(row['title_non_stop_words']) + [0] * len(row['content_non_stop_words']),
+        axis=1)
+
+
+def _add_number_of_non_stop_words(data):
+    split_row = lambda row: [x for x in row.split() if x not in stop_words]
+    data['title_non_stop_words'] = data['title'].apply(split_row)
+    data['content_non_stop_words'] = data['content'].apply(split_row)
+
+@model.FeatureManager
+def _times_word_in(data, column):
+    return data.apply(
+        lambda row: [row[column].split().count(word)
+                     for word in row['titlecontent'].split()
+                     if word not in stop_words],
+        axis=1)
+
+class Features(model.Features):
     def __init__(self, functional_test=False, changes=False):
         self.functional_test = functional_test
         self.tf_idf_vectorizer = None
         self.changes = changes
+        self.save = False
 
     def fit(self, train_data):
         self._train_tf_idf_vectorizer(train_data)
 
     def transform(self, data):
-        self._add_number_of_non_stop_words(data)
-        features = [
-                self._get_tf_idf_features_per_word(data),
-                self._times_word_in(data, 'title'),
-                self._times_word_in(data, 'content'),
-                self._is_in_question(data),
-        ]
-     #   if self.changes:
-        features += [self._title_or_content(data)]
+        features = self._get_features_per_row(data)
+        for i, feat in enumerate(features):
+            features[i] = list(itertools.chain(*feat))
```

Comment:
Well, this line is classified as an addition in your changeset, so I doubt it would be overwritten by changes from my branch. It's a detail but if we want to use `from_iterable`, I think you should already make the change.